tfhe/src/core_crypto/gpu/mod.rs

```diff
@@ -152,6 +176,63 @@ impl CudaStream {
         }
     }
 
+    /// Copy data between two arrays in the GPU, where the destination is allocated with
+    /// less memory space than the source
+    ///
+    /// # Safety
+    ///
+    /// - `src` __must__ be a valid pointer to the GPU global memory
+    /// - `dest` __must__ be a valid pointer to the GPU global memory
+    /// - [CudaStream::synchronize] __must__ be called after the copy
+    /// as soon as synchronization is required
+    pub unsafe fn copy_gpu_to_gpu_smaller_dest_async<T>(
+        &self,
+        dest: &mut CudaVec<T>,
+        src: &CudaVec<T>,
+        shift: u32,
+    ) where
+        T: Numeric,
+    {
+        let src_shifted_size = (src.len() + shift as usize) * std::mem::size_of::<T>();
+        let src_ptr = src.as_c_ptr().add(shift as usize);
+        let dest_size = dest.len() * std::mem::size_of::<T>();
+        assert!(src_shifted_size >= dest_size);
+        // We check that src is not empty to avoid invalid pointers
+        if src_shifted_size > 0 {
+            cuda_memcpy_async_gpu_to_gpu(
+                dest.as_mut_c_ptr(),
+                src_ptr,
+                dest_size as u64,
+                self.as_c_ptr(),
+            );
+        }
+    }
+
+    /// Copies data between different arrays in the GPU with a ptr shift
+    ///
+    /// # Safety
+    ///
+    /// - `src` __must__ be a valid pointer to the GPU global memory
+    /// - `dest` __must__ be a valid pointer to the GPU global memory
+    /// - [CudaStream::synchronize] __must__ be called after the copy
+    /// as soon as synchronization is required
+    pub unsafe fn copy_gpu_to_gpu_with_ptr_shift_async<T>(
+        &self,
+        dest: &mut CudaVec<T>,
+        src: &CudaVec<T>,
+        shift: u32,
+    ) where
+        T: Numeric,
+    {
+        let dest_ptr = dest.as_mut_c_ptr().add(shift as usize);
+        assert!(dest.len() - shift as usize >= src.len());
+        let size = src.len() * std::mem::size_of::<T>();
+        // We check that src is not empty to avoid invalid pointers
+        if size > 0 {
+            cuda_memcpy_async_gpu_to_gpu(dest_ptr, src.as_c_ptr(), size as u64, self.as_c_ptr());
+        }
+    }
+
```

Comment:
Same comments as for `memset_with_ptr_shift_async`

I thing these should be move as method of cuda vec, and use `RangeBouds` to provide a better API.


```rust
impl CudaVec<T> {
	pub unsafe fn copy_range_from_gpu_async<R>(&mut self, range: R, src: &CudaVec<T>) {
		// share code with `memset_with_ptr_shift_async` to get the start and end from a RangeBounds
		let (start, end) = self.range_bounds_to_start_end();

		// check start_end is correct for both self, and src
		// call memcpy
	}
}
```

usage would be `cuda_vec1.copy_range_from_gpu_async(23.., &cuda_vec_2, stream)`